### PR TITLE
[Original, draft] Add Elrond coin

### LINKED ISSUE #444: Adopt Zilliqa to bech32 address format

### PR DESCRIPTION
**Pull request by andreibancioiu**
## Description

Extended Wallet-core with blockchain & coin **Elrond**.

## Testing instructions

 * interacting with `walletconsole`
 * running the tests

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [ ] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [ ] If there is a related Issue, mention it in the description (e.g. 'Fixes #444 ).

## New blockchain checklist

 - [x] Add the coin definition to `coins.json`.
 - [x] Add coin dispatcher to `src/Coin.cpp`
 - [x] Create tests in `tests/Elrond/TWCoinTypeTests.cpp`
 - [x] Implement `Entry.cpp`
 - [x] Implement `Address.cpp`
 - [ ] ~~Implement `Transaction.cpp` if necessary~~. *Was not necessary*.
 - [x] Implement `Signer.cpp`
 - [x] Write test: mnemonic phrase - > Address derivation test. Put this test in ~~`CoinTests.cpp`~~ `TWHDWalletTests.cpp`.
 - [x] Transaction signing tests, at least one mainnet transaction test.
 - [ ] ~~Add stake, unstake, get rewards tests if the blockchain is PoS like.~~ *Will be added later in time.*
 - [ ] ~~Implement C interface in `src/interface`~~ *Not applicable*.
 - [x] Add tests for `TWAnyAddress` 
 - [x] Add tests for  `TWAnySigner`
 - [x] Add tests in `ElrondTests.swift`
 - [x] Validate generated code in Android project. Write integration tests. *Maiar Android Wallet*.
 - [x] Validate generated code in iOS project. Write integration tests. *Maiar iOS Wallet*.
 - [x] Extend test: `CoinAddressDerivationTests.cpp` 
 - [x] Extend test: `CoinAddressValidationTests.cpp`
 - [x] Extend test: `TWHRPTests.cpp`
 - [x] Extend test `CoinAddressDerivationTests.kt`
 - [x] Extend test `CoinAddressDerivationTests.swift`
 - [ ] Upload coin icon to `trustwallet/assets` if necessary.

### CODE REVIEW

File: include/TrustWalletCore/TWCoinType.h [hunk 1]
```diff
@@ -79,6 +79,7 @@ enum TWCoinType {
     TWCoinTypeKusama = 434,
     TWCoinTypePolkadot = 354,
     TWCoinTypeFilecoin = 461,
+    TWCoinTypeElrond = 2003,
```
Comment (AdoAdoAdo) [inline include/TrustWalletCore/TWCoinType.h]: update the coin type

File: tests/Elrond/TestAccounts.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+const auto ALICE_BECH32 = "erd1l453hd0gt5gzdp7czpuall8ggt2dcv5zwmfdf3sd3lguxseux2fsmsgldz";
+const auto ALICE_SEED_HEX = "1a927e2af5306a9bb2ea777f73e06ecc0ac9aaa72fb4ea3fecf659451394cccf";
+const auto ALICE_PUBKEY_HEX = "fd691bb5e85d102687d81079dffce842d4dc328276d2d4c60d8fd1c3433c3293";
+const auto BOB_BECH32 = "erd1cux02zersde0l7hhklzhywcxk4u9n4py5tdxyx7vrvhnza2r4gmq4vw35r";
+const auto BOB_SEED_HEX = "e3a3a3d1ac40d42d8fd4c569a9749b65a1250dd3d10b6f4e438297662ea4850e";
+const auto BOB_PUBKEY_HEX = "c70cf50b238372fffaf7b7c5723b06b57859d424a2da621bcc1b2f317543aa36";
+const auto CAROL_BECH32 = "erd19nu5t7hszckwah5nlcadmk5rlchtugzplznskffpwecygcu0520s9tnyy0";
+const auto CAROL_SEED_HEX = "a926316cc3daf8ff25ba3e417797e6dfd51f62ae735ab07148234732f7314052";
+const auto CAROL_PUBKEY_HEX = "2cf945faf0162ceede93fe3addda83fe2ebe2041f8a70b2521767044638fa29f";
```
Comment (AdoAdoAdo) [inline tests/Elrond/TestAccounts.h]: newline at eof

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/elrond/TestElrondAddress.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Test
+    fun testAddressFromPrivateKey() {
+        val key = PrivateKey(aliceSeedHex.toHexByteArray())
+        val pubKey = key.publicKeyEd25519
+        val address = AnyAddress(pubKey, CoinType.ELROND)
+        val expected = AnyAddress(aliceBech32, CoinType.ELROND)
+
+        assertEquals(alicePubKeyHex, pubKey.data().toHex())
+        assertEquals(expected.description(), address.description())
```
Comment (AdoAdoAdo) [inline android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/elrond/TestElrondAddress.kt]: I guess here aliceBech32 should be identical to expected.description()
is there a need to create "expected"?

File: coins.json [hunk 4]
```diff
@@ -1330,5 +1330,27 @@
...
             "clientDocs": "https://docs.lotu.sh"
         }
+    },
+    {
+        "id": "elrond",
+        "name": "Elrond",
+        "symbol": "ERD",
+        "decimals": 18,
+        "blockchain": "ElrondNetwork",
+        "derivationPath": "m/44'/2003'/0'",
```
Comment (AdoAdoAdo) [inline coins.json]: update derivation path

File: src/Elrond/Address.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    /// Determines whether a string makes a valid address.
+    static bool isValid(const std::string& string);
+
+    Address() : Bech32Address(hrp) {}
+
+    /// Initializes an address with a key hash.
+    Address(Data keyHash) : Bech32Address(hrp, keyHash) {}
+
+    /// Initializes an address with a public key.
+    Address(const PublicKey& publicKey) : Bech32Address(hrp, publicKey.bytes) {}
```
Comment (AdoAdoAdo) [inline src/Elrond/Address.h:31]: is this correct?
don't we need a hasher and call Bech32Address(hrp, *hasher*, publicKey) ?

File: src/Elrond/Signer.cpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+#include <google/protobuf/util/json_util.h>
+
+using namespace TW;
+using namespace TW::Elrond;
+
+
+Proto::SigningOutput Signer::sign(const Proto::SigningInput &input) noexcept {
+    auto privateKey = PrivateKey(input.private_key());
+    auto signableAsString = Elrond::serializeTransactionToSignableString(input.transaction());
```
Comment (AdoAdoAdo) [inline src/Elrond/Signer.cpp]: I think Elrond qualifier not needed here as it is already in scope

File: src/Elrond/Signer.cpp [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    auto signableAsData = TW::data(signableAsString);
+    auto signature = privateKey.sign(signableAsData, TWCurveED25519);
+    auto encodedSignature = hex(signature);
+    auto serializedTransaction = Elrond::serializeSignedTransaction(input.transaction(), encodedSignature);
```
Comment (AdoAdoAdo) [inline src/Elrond/Signer.cpp]: I think Elrond qualifier not needed here as it is already in scope

File: src/Elrond/Signer.cpp [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    protoOutput.set_signature(encodedSignature);
+    protoOutput.set_signed_transaction(serializedTransaction);
+    return protoOutput;
+}
+
+std::string Signer::signJSON(const std::string& json, const Data& key) {
+    auto input = Proto::SigningInput();
+    google::protobuf::util::JsonStringToMessage(json, &input);
+    input.set_private_key(key.data(), key.size());
+    auto output = Signer::sign(input);
```
Comment (AdoAdoAdo) [inline src/Elrond/Signer.cpp]: I think Signer qualifier is not needed

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: tests/Bech32AddressTests.cpp [hunk 9]
```diff
@@ -48,6 +52,10 @@ TEST(Bech32Address, Invalid) {
     ASSERT_FALSE(Bech32Address::isValid(""));
     ASSERT_FALSE(Bech32Address::isValid("0x"));
     ASSERT_FALSE(Bech32Address::isValid("91cddcebe846ce4d47712287eee53cf17c2cfb7"));
+
+    ASSERT_FALSE(Bech32Address::isValid("", "erd"));
+    ASSERT_FALSE(Bech32Address::isValid("erd1cux02zersde0l7hhklzhywcxk4u9n4py5tdxyx7vrvhnza2r4gmq4vw35!", "erd"));
+    ASSERT_FALSE(Bech32Address::isValid("terd19nu5t7hszckwah5nlcadmk5rlchtugzplznskffpwecygcu0520s9tnyy0", "erd"));
```
Comment (camilbancioiu) [inline tests/Bech32AddressTests.cpp]: xerd

File: tests/CoinAddressValidationTests.cpp [hunk 10]
```diff
@@ -373,4 +373,10 @@ TEST(Coin, ValidateAddressVeChain) {
     EXPECT_EQ(normalizeAddress(TWCoinTypeVeChain, "0x9d8a62f656a8d1615c1294fd71e9cfb3e4855a4f"), "0x9d8A62f656a8d1615C1294fd71e9CFb3E4855A4F");
 }
 
+TEST(Coin, ValidateAddressElrond) {
+    EXPECT_TRUE(validateAddress(TWCoinTypeElrond, "erd1l453hd0gt5gzdp7czpuall8ggt2dcv5zwmfdf3sd3lguxseux2fsmsgldz"));
+    EXPECT_FALSE(validateAddress(TWCoinTypeElrond, "terd1l453hd0gt5gzdp7czpuall8ggt2dcv5zwmfdf3sd3lguxseux2fsmsgldz"));
```
Comment (camilbancioiu) [inline tests/CoinAddressValidationTests.cpp]: xerd

File: tests/Elrond/AddressTests.cpp [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    ASSERT_TRUE(Address::isValid(ALICE_BECH32));
+    ASSERT_TRUE(Address::isValid(BOB_BECH32));
+    ASSERT_TRUE(Address::isValid(CAROL_BECH32));
+}
+
+TEST(ElrondAddress, Invalid) {
+    ASSERT_FALSE(Address::isValid(""));
+    ASSERT_FALSE(Address::isValid("foo"));
+    ASSERT_FALSE(Address::isValid("10z9xdugayn528ksaesdwlhf006fw5sg2qmmm0h52fvxczwgesyvq5pwemr"));
+    ASSERT_FALSE(Address::isValid("nerd10z9xdugayn528ksaesdwlhf006fw5sg2qmmm0h52fvxczwgesyvq5pwemr"));
```
Comment (camilbancioiu) [inline tests/Elrond/AddressTests.cpp]: xerd

**Review by camilbancioiu — CHANGES_REQUESTED** [review]

File: tests/Elrond/TestAccounts.h (hunk 2 above)
Comment (andreibancioiu) [reply tests/Elrond/TestAccounts.h]: Fixed.

File: src/Elrond/Address.h (hunk 5 above)
Comment (andreibancioiu) [reply src/Elrond/Address.h:31]: As discussed, we use `bech32(pubKey)`, not `bech32(hash(pubKey))`.

File: src/Elrond/Signer.cpp (hunk 6 above)
Comment (andreibancioiu) [reply src/Elrond/Signer.cpp]: Fixed.

File: src/Elrond/Signer.cpp (hunk 8 above)
Comment (andreibancioiu) [reply src/Elrond/Signer.cpp]: Fixed.

File: tests/Bech32AddressTests.cpp (hunk 9 above)
Comment (andreibancioiu) [reply tests/Bech32AddressTests.cpp]: Fixed.

File: tests/CoinAddressValidationTests.cpp (hunk 10 above)
Comment (andreibancioiu) [reply tests/CoinAddressValidationTests.cpp]: Fixed.

File: tests/Elrond/AddressTests.cpp (hunk 11 above)
Comment (andreibancioiu) [reply tests/Elrond/AddressTests.cpp]: Fixed.

File: src/Elrond/Signer.cpp (hunk 7 above)
Comment (andreibancioiu) [reply src/Elrond/Signer.cpp]: Fixed.

File: include/TrustWalletCore/TWCoinType.h (hunk 1 above)
Comment (andreibancioiu) [reply include/TrustWalletCore/TWCoinType.h]: Fixed.

File: coins.json (hunk 4 above)
Comment (andreibancioiu) [reply coins.json]: Fixed.

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/elrond/TestElrondAddress.kt (hunk 3 above)
Comment (andreibancioiu) [reply android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/elrond/TestElrondAddress.kt]: Fixed.

**Review by camilbancioiu — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Comment by andreibancioiu** [comment]
We can close this PR, the official one is against the parent repository.

**Pull request closed without merging**